openqa-post-similarity

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/python
+
+import os
+import Levenshtein
+from http import client
+from openqa_client.client import OpenQA_Client
+import json
+import argparse
+import logging
+from tqdm import tqdm
+import sys
+
+"""
+This script scans all autoinst-log.txt files in the testresults directory searching
+for error messages. Then it computes the Levenshtein distance and posts the
+results as openQA comments.
+"""
+
+id_msg = {}
+result = {}
+
+
+# I used a JSON file to store some intermediate computation results to reduce time cost.
+# This part finds error messages in autoinst-log.txt
+# Maybe there are multiple errors in a single autoinst-log.txt
+# I think some different code may trigger the same bug.
+# The first line is the most important, so I focus on the first lines of the errors.
+def log_set():
+    logger = logging.getLogger()
+    logger.setLevel(logging.INFO)
+    logger.addHandler(logging.StreamHandler(sys.stdout))
+    return logger
+
+
+def read_errors(path):
+    global id_msg
+    with open(path, 'r') as f:
+        text = f.read()
+        anchor = 0
+        begin = text.find("Test died", anchor)
+        while begin != -1:
+            anchor = begin + 1
+            end = text.find("\n", begin)
+            id_msg[path.split("/")[-2]] = text[begin:end]
+            begin = text.find("Test died", anchor)
+
+
+def read_id_msg(logger, testdir):
+    global id_msg
+    if not os.path.exists("id_msg.json"):
+        logger.info("id_msg.json does not exist.")
+        all_dirs = os.listdir(testdir)
+        all_dirs = [testdir + item for item in all_dirs]
+        logs = [item + '/autoinst-log.txt' for item in all_dirs]
+
+        for path in logs:
+            read_errors(path)
+        id_msg_json = json.dumps(id_msg, sort_keys=False, indent=4, separators=(',', ':'))
+        with open("id_msg.json", "w") as f:
+            f.write(id_msg_json)
+
+    # If there exists the JSON file
+    else:
+        logger.info("id_msg.json exists.")
+        with open("id_msg.json", "r") as f:
+            id_msg = json.load(f)
+        all_dirs = os.listdir(testdir)
+        checked_dirs = list(id_msg.keys())
+        unchecked_dirs = [i for i in all_dirs if i not in checked_dirs]
+        for job_id in unchecked_dirs:
+            path = testdir + job_id + "/autoinst-log.txt"
+            read_errors(path)
+        id_msg_json = json.dumps(id_msg, sort_keys=False, indent=4, separators=(',', ':'))
+        with open("id_msg.json", "w") as f:
+            f.write(id_msg_json)
```

Comment:
I still think there is lots of duplicated code here in the if and else block. The last 3 lines are exactly the same, and it feels like the rest of the code could also be shared if it was rewritten.
But I think it's ok for now, as we are not using it in production yet.

Reply:
Thank you. I will rewrite it as soon as possible.